### ISSUE TITLE
Default repair strategy is inefficient when repairing multiple rules

### ISSUE THREAD

**Issue by slarse** [SpoonLabs/sorald#172]
The default repair strategy is highly inefficient when repairing multiple rules. In short, it parses the input file, scans it for one rule, repairs any violations, and then prints to an intermediate file. For the next rule, it parses the intermediate file, and so on. I don't know the original reasoning for this approach. The whole thing can be found here: https://github.com/SpoonLabs/sorald/blob/c76d525603345ad82149951ce3f2c7339ce6236d/src/main/java/sorald/DefaultRepair.java#L31-L88

This incurs a heavy performance penalty when processing multiple rules, as parsing is heavy with Spoon, and actually writing content to a file is an inherently slow operation. In essence, when processing multiple rules, most of the running time is likely spent parsing and printing.

In theory, this printing to an intermediate file should not be necessary, as we _should_ have `parse(print(ast)) == ast`. Of course, practice is not theory. In particular the sniper printer is sensitive to transformations, and may cause issues if we try to apply too many at the same time. From a theoretical perspective, the end goal should however be to apply all transformations on the AST, without printing and reparsing thrown into the mix.

**In practice, this may not be a problem.** It is reasonable to assume that, when repairing rules with Sorald, developers will want to fix only one or perhaps a few closely related rules at a time. We however need some data to support this assumption.

To summarize, repairing multiple rules with Sorald is currently fairly slow going, and writing to disk and reparsing is very likely the bottleneck. Although it should be noted that I haven't actually done any profiling, there's a slim chance that I'm wrong. But I don't think so. :)
